Ability to add mood events.

mauliksamit opened the pull request:
Changes:
UI and logic for adding fields of mood events

resolves #41
resolves #42
resolves #89
resolves #116

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/model/MoodEvent.java:
nit: reformat required

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,181 @@
+        // Set up repositories and location client
+        moodEventRepository = new MoodEventRepository();
+        participantRepository = new ParticipantRepository();
+        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
```
This implementation has already been written in `LocationHandler` we don't need duplicate impl. Please use that.

If you need some implementation that is missing in the class feel free to add it there, So that everyone can use it rather than copying it everywhere.

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,181 @@
+
+        // Get selected emotional state (required)
+        MoodEvent.EmotionalState emotionalState = (MoodEvent.EmotionalState) emotionalStateSpinner.getSelectedItem();
+        if (emotionalState == null) {
```
Also add log statements so we know what really happened when we see an error. See `LoginPage` for example. Applies to whole file.

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,181 @@
+        // Handle location (optional, based on checkbox)
+        AtomicReference<Map<String, Object>> geoInfo = new AtomicReference<>();
+        if (locationCheckbox.isChecked()) {
+            if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, LOCATION_PERMISSION_REQUEST_CODE);
+                Toast.makeText(this, "Location permission required. Please grant permission and try again.", Toast.LENGTH_SHORT).show();
+                return;
+            }
```
Again, use `LocationHandler` duplicate impl.

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,181 @@
+                                           MoodEvent.EmotionalState emotionalState, MoodEvent.SocialSituation socialSituation,
+                                           String trigger) {
+        // Fetch the current participant from Firebase using ParticipantRepository
+        participantRepository.fetchParticipant(getCurrentUsername(), participant -> {
```
You don't need to fetchParticipant to get the DocumentReference. Use the available public method in ParticipantRepository

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,181 @@
+
+            // Get the participant's username or use a unique identifier (e.g., document ID)
+            String username = participant.getUsername();
+            DocumentReference participantRef = FirebaseFirestore.getInstance().collection("participants").document(username);
```
Again, refrain from using firebase direct. Use the already available static `FirebaseService` written if you need to use it. Otherwise always use ParticipantRepository as mush as possible If you need a method that doesn't exist in it "Add it".

ArmaanKatyal reviewed (changes requested):
Some fixing required with diuplicate impl.

mauliksamit commented:
Changes:
UI and logic for adding fields of mood events.

Redundancies addressed by using ParticipationRepository,java and LocationHandler.java
Added Validation Checks.

resolves #41
resolves #42
resolves #44
resolves #81
resolves #82
resolves #89
resolves #115
resolves #116

mauliksamit commented:
Changes were made to the files listed below.

MoodEventArrayAdapter.java
HomePage.java
MoodEvents.java

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,417 @@
+package com.example.bread.view;
+
+import android.os.Bundle;
+import android.util.Log; // Added for logging
```
nit: unnecessary comment

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/HomePage.java:
```diff
@@ -38,14 +38,18 @@ protected void onCreate(Bundle savedInstanceState) {
                 replaceFragment(new MapFragment());
             } else if (itemId == R.id.add) {
-                replaceFragment(new AddFragment());
+                // Start AddMoodEventActivity instead of replacing with a fragment
+                Intent intent = new Intent(HomePage.this, AddMoodEventActivity.class);
+                startActivity(intent);
+                // Return false to prevent the navigation item from staying selected (optional)
+                return false;
```
For now its fine, but it completely changes the UI. We would need to move to a fragment as the navigation bar disappears when the user launches this activity which is not good.

mauliksamit replied:
yup agreed

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/HomePage.java:
```diff
@@ -1,6 +1,7 @@
 package com.example.bread.view;
 
 import android.annotation.SuppressLint;
+import android.content.Intent; // Added for Intent
```
nit: comment not required

ArmaanKatyal commented on code/app/src/main/res/drawable/rounded_background.xml:
Do we want to check the background of our app static or changeable depending on user selection for light or dark mode.

animeshm19 replied:
we are not doing light and dark mode anywhere in the app, so Idts we need to worry about that

ArmaanKatyal commented on code/app/src/main/res/layout/activity_add_mood_event.xml:
```diff
@@ -0,0 +1,189 @@
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="Date"
+                    android:textColor="#FFFFFF"
+                    android:paddingBottom="4dp"/>
+
+                <EditText
+                    android:id="@+id/dateEditText"
+                    android:layout_width="match_parent"
+                    android:layout_height="40dp"
+                    android:background="@drawable/rounded_background"
+                    android:enabled="false"
+                    android:textColor="#FFFFFF"
+                    android:padding="4dp"/>
+            </LinearLayout>
+
+            <LinearLayout
+                android:layout_width="0dp"
+                android:layout_height="wrap_content"
+                android:layout_weight="1"
+                android:orientation="vertical"
+                android:paddingLeft="16dp">
+
+                <TextView
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="Time"
+                    android:textColor="#FFFFFF"
+                    android:paddingBottom="4dp"/>
+
+                <EditText
+                    android:id="@+id/timeEditText"
+                    android:layout_width="match_parent"
+                    android:layout_height="40dp"
+                    android:background="@drawable/rounded_background"
+                    android:enabled="false"
+                    android:textColor="#FFFFFF"
+                    android:padding="4dp"/>
+            </LinearLayout>
+        </LinearLayout>
```
User has no control over date and time, They are automatically populated by firebase. So we need to remove these fields from here.

mauliksamit replied:
👍

ArmaanKatyal reviewed (commented):
Needs quite a few changes. Plus UI testing is required for this PR. So can you please also add that. Thanks!

mauliksamit commented on code/app/src/main/java/com/example/bread/model/MoodEvent.java:
```diff
@@ -29,21 +30,21 @@ public enum EmotionalState {
      * Enum representing the different social situations a user can be in
      */
     public enum SocialSituation {
+        NONE,
```
Just so that the default selection for social situation is set to none instead of some other option.

mauliksamit commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,417 @@
+        moodEvent.setTrigger(trigger);
+        moodEvent.setAttachedImage(null); // TODO: Implement image upload functionality
+        Log.d(TAG, "MoodEvent created: " + moodEvent.toString());
+        Log.d(TAG, "Timestamp (before save): " + (moodEvent.getTimestamp() != null ? moodEvent.getTimestamp().toString() : "null (to be set by server)"));
```
was useful for debugging purposes

mauliksamit commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,417 @@
+                Log.i(TAG, "Location attached to mood event: " + geoInfo);
+            } catch (Exception e) {
+                Log.e(TAG, "Error generating geo info: " + e.getMessage(), e);
+                Toast.makeText(this, "Error attaching location: " + e.getMessage(), Toast.LENGTH_SHORT).show();
```
got it

mauliksamit commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,417 @@
+        locationHandler.stopLocationUpdates();
+    }
+}
+//package com.example.bread.view;
```
👍

mauliksamit commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,417 @@
+                },
+                e -> {
+                    Log.e(TAG, "Failed to save mood event: " + e.getMessage(), e);
+                    Toast.makeText(this, "Failed to save mood: " + e.getMessage(), Toast.LENGTH_SHORT).show();
```
will be removed

mauliksamit commented on code/app/src/main/java/com/example/bread/model/MoodEvent.java:
```diff
@@ -29,21 +30,21 @@ public enum EmotionalState {
         WITH_COWORKERS,
         WITH_STRANGERS,
-        NONE;
+        ;
```
will be fixed

mauliksamit commented on code/app/src/main/java/com/example/bread/controller/MoodEventArrayAdapter.java:
```diff
@@ -62,51 +62,92 @@ public View getView(int position, @Nullable View convertView, @NonNull ViewGroup
+            Log.d(TAG, "Current user display name: " + participantUsername);
+        } else {
+            Log.w(TAG, "No current Firebase user found");
+            participantUsername = "Unknown";
```
got it

mauliksamit commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -0,0 +1,417 @@
+            isValid = false;
+        } else {
+            Log.d(TAG, "Event title validated: " + eventTitle);
+            eventTitleEditText.setError(null);
```
just in case it needs to clear an existing error state from previous validations

animeshm19 commented on code/app/src/main/res/layout/activity_add_mood_event.xml:
```diff
@@ -0,0 +1,189 @@
+            android:layout_marginTop="16dp"/>
+
+    </LinearLayout>
+
```
It's a low priority, but add start & end padding on the hint. Also, reduce the corner of the UI for all the fields since it's almost hitting the screen from both ends (left, right).

mauliksamit commented:
UI and logic for adding fields of mood events.

Changes: UI improvements, removed Date/Field input fields for users, fixed bug in HistoryFragment.java which was causing the app to crash

resolves https://github.com/cmput301-w25/project-bread/issues/41
resolves https://github.com/cmput301-w25/project-bread/issues/42
resolves https://github.com/cmput301-w25/project-bread/issues/44
resolves https://github.com/cmput301-w25/project-bread/issues/81
resolves https://github.com/cmput301-w25/project-bread/issues/82
resolves https://github.com/cmput301-w25/project-bread/issues/89
resolves https://github.com/cmput301-w25/project-bread/issues/115
resolves https://github.com/cmput301-w25/project-bread/issues/116
resolves #154

mauliksamit commented:
UI and logic for adding fields of mood events.

Changes: UI improvements, removed Date/Field input fields for users, fixed bug in HistoryFragment.java which was causing the app to crash

resolves https://github.com/cmput301-w25/project-bread/issues/41
resolves https://github.com/cmput301-w25/project-bread/issues/42
resolves https://github.com/cmput301-w25/project-bread/issues/44
resolves https://github.com/cmput301-w25/project-bread/issues/81
resolves https://github.com/cmput301-w25/project-bread/issues/82
resolves https://github.com/cmput301-w25/project-bread/issues/89
resolves https://github.com/cmput301-w25/project-bread/issues/115
resolves https://github.com/cmput301-w25/project-bread/issues/116
resolves https://github.com/cmput301-w25/project-bread/issues/154

UI tests will be added soon

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/fragment/HistoryFragment.java:
```diff
@@ -102,7 +102,10 @@ private void loadMoodEvents() {
                     if (moodEvents != null) {
                         moodEventArrayList.clear();
-                        moodEventArrayList.addAll(moodEvents);
+                        //moodEventArrayList.addAll(moodEvents);
```
It’s best not to filter out events that lack a timestamp, since doing so would remove them from the list. Instead, let’s update the compareTo method to handle missing timestamps the same way you did before (also checkout PR #152). Sorry for the late change, but after reviewing the documentation, it appears that ServerTimestamp isn’t always set immediately. As a result, we may need to store two timestamps—one from the app and one from the server. If the server timestamp is missing, we can default to the app’s timestamp. This should be our final fix. Thanks!

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java:
```diff
@@ -124,7 +123,7 @@ private void saveMoodEvent() {
         } else {
             Log.d(TAG, "Event title validated: " + eventTitle);
-            eventTitleEditText.setError(null);
+            //eventTitleEditText.setError(null);
```
Is this by mistake?

ArmaanKatyal reviewed (approved):
Sorry for the delay, but this looks great now! Just add the tests, and you’re good to merge everything in. Thanks!

ArmaanKatyal commented on code/app/src/main/java/com/example/bread/model/MoodEvent.java:
```diff
@@ -190,8 +202,21 @@ public Map<String, Object> generateGeoInfo(Location location) {
+        }
+        if (event.getTimestamp() == null) {
+            return -1; // Non-null comes before null
+        }
```
Just put this back in. but checkout PR #152 which also adds the same thing.
